sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update_slm.hpp

```diff
@@ -0,0 +1,720 @@
+            // arg0 and arg1 must provide values which guarantee the loop
+            // is not endless:
+            if constexpr (UseMask) {
+              for (simd<T, N> old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val, m);
+                   any(old_val < exp_val, !m);
+                   old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val, m))
+                ;
+            } else {
+              for (simd<T, N> old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val);
+                   any(old_val < exp_val, !m);
+                   old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val))
+                ;
+            }
+          }
+        }
+        // TODO: replace this API with unified version.
+        auto data0 = lsc_slm_gather<T>(slm_offsets);
+        data0.copy_to(arr);
+      });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(arr, q);
+    return false;
+  }
+  int err_cnt = 0;
+
+  for (int i = 0; i < size; ++i) {
+    T gold = ImplF<T, N>::gold(i, UseMask);
+    T test = arr[i];
+
+    if ((gold != test) && (++err_cnt < 10)) {
+      if (err_cnt == 1) {
+        std::cout << "\n";
+      }
+      std::cout << "  failed at index " << i << ": " << test << " != " << gold
+                << "(gold)\n";
+    }
+  }
+  if (err_cnt > 0) {
+    std::cout << "  FAILED\n  pass rate: "
+              << ((float)(size - err_cnt) / (float)size) * 100.0f << "% ("
+              << (size - err_cnt) << "/" << size << ")\n";
+  } else {
+    std::cout << " passed\n";
+  }
+  free(arr, q);
+  return err_cnt == 0;
+}
+
+template <class T, int N, template <class, int> class ImplF, bool UseMask>
+bool test_slm_acc(queue q) {
+  constexpr auto op = ImplF<T, N>::atomic_op;
+  using CurAtomicOpT = decltype(op);
+  constexpr int n_args = ImplF<T, N>::n_args;
+
+  std::cout << "SLM ACC testing" << " op=" << to_string(op)
+            << " T=" << esimd_test::type_name<T>() << " N=" << N << "\n\t"
+            << " UseMask=" << (UseMask ? "true" : "false")
+            << "{ thr_per_group=" << threads_per_group
+            << " n_groups=" << n_groups << " start_ind=" << start_ind
+            << " masked_lane=" << masked_lane << " repeat=" << repeat
+            << " stride=" << stride << " }...";
+
+  constexpr size_t size = start_ind + (N - 1) * stride + 1;
+  T *arr = malloc_shared<T>(size, q);
+  constexpr int n_threads = threads_per_group * n_groups;
+
+  for (int i = 0; i < size; ++i) {
+    arr[i] = ImplF<T, N>::init(i);
+  }
+
+  range<1> glob_rng(n_threads);
+  range<1> loc_rng(threads_per_group);
+  nd_range<1> rng(glob_rng, loc_rng);
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      local_accessor<T, 1> LocalAcc(threads_per_group * N, cgh);
```

Comment:
I think this line allocates too few memory.
The writes/loads to SLM will use stride > 1. Thus need more memory.

Reply:
Done.